libraries/node-core-library/src/FileSystem.ts

```diff
@@ -304,21 +402,55 @@ export class FileSystem {
    */
   public static readFolder(folderPath: string, options?: IFileSystemReadFolderOptions): string[] {
     options = {
-      absolutePaths: false,
+      ...READ_FOLDER_DEFAULT_OPTIONS,
       ...options
     };
 
-    if (!FileSystem.exists(folderPath)) {
-      throw new Error(`Folder does not exist: "${folderPath}"`);
+    let fileNames: string[]
+    try {
+      // @todo: Update this to use Node 10's `withFileTypes: true` option when we drop support for Node 8
+      fileNames = fsx.readdirSync(folderPath);
+    } catch (e) {
+      if (FileSystem.isNotExistError(e)) {
+        throw new Error(`Folder does not exist: "${folderPath}"`);
+      } else {
+        throw e;
+      }
     }
 
-    const fileNames: string[] = fsx.readdirSync(folderPath);
-
     if (options.absolutePaths) {
-      return fileNames.map(fileName => pathUtilities.resolve(folderPath, fileName));
+      return FileSystem._resolvePaths(fileNames, folderPath);
+    } else {
+      return fileNames;
+    }
+  }
+
+  /**
+   * An async version of {@link FileSystem.readFolder}.
+   */
+  public static async readFolderAsync(folderPath: string, options?: IFileSystemReadFolderOptions): Promise<string[]> {
+    options = {
+      ...READ_FOLDER_DEFAULT_OPTIONS,
+      ...options
+    };
+
+    let fileNames: string[];
+    try {
+      // @todo: Update this to use Node 10's `withFileTypes: true` option when we drop support for Node 8
+      fileNames = await fsx.readdir(folderPath);
+    } catch (e) {
+      if (FileSystem.isNotExistError(e)) {
+        throw new Error(`Folder does not exist: "${folderPath}"`);
+      } else {
+        throw e;
+      }
     }
 
-    return fileNames;
+    if (options.absolutePaths) {
+      return FileSystem._resolvePaths(fileNames, folderPath);
```

Comment:
Why do we apply `async` to operations like `fsx.stat()` that are near-instantaneous? It must be a goal of providing near  realtime scheduling guarantees, or perhaps a fear of some API call taking too long and stalling the entire app.

But `FileSystem._resolvePaths()` involves disk access. Why isn't it async? Where should we draw the line with this paranoia. :-)

Reply:
>Why do we apply `async` to operations like `fsx.stat()` that are near-instantaneous?

They can still block. For instance, if the path is on a very slow network-connected resource.

>But `FileSystem._resolvePaths()` involves disk access.

No it doesn't - it's purely string manipulation: [posix](https://github.com/nodejs/node/blob/master/lib/path.js#L973-L1002), [windows](https://github.com/nodejs/node/blob/master/lib/path.js#L130-L263)